### PR TITLE
Rock config

### PR DESCRIPTION
**Pull request by joakim-hove**
Internalize rock pecularities

Dwonstream: https://github.com/OPM/opm-simulators/pull/2303

### CODE REVIEW

**Comment by bska** [comment]
Is this intended, at least in part, to address the problem raised in Issue #1357?

**Comment by joakim-hove** [comment]
> Is this intended, at least in part, to address the problem raised in Issue #1357?

The main purpose is to reduce `Deck` usage downstream, but #1357 should be addressed along the way.

**Comment by joakim-hove** [comment]
jenkins build this opm-simulators=2302 please

**Comment by joakim-hove** [comment]
jenkins build this opm-simulators=2303 please

**Comment by joakim-hove** [comment]
jenkins build this opm-simulators=2303 please

**Comment by joakim-hove** [comment]
jenkins build this opm-simulators=2303 please

**Comment by joakim-hove** [comment]
jenkins build this opm-simulators=2303 please

**Comment by joakim-hove** [comment]
jenkins build this opm-simulators=2303 please

**Comment by joakim-hove** [comment]
jenkins build this opm-simulators=2303 please

**Comment by joakim-hove** [comment]
jenkins build this opm-simulators=2303 please

**Comment by joakim-hove** [comment]
jenkins build this opm-simulators=2303 please

**Comment by joakim-hove** [comment]
jenkins build this opm-simulators=2303 please

**Comment by joakim-hove** [comment]
Comments? I'll merge later today if not.

File: src/opm/parser/eclipse/EclipseState/SimulationConfig/RockConfig.cpp [hunk 1]
```diff
@@ -0,0 +1,157 @@
...
+        return "SATNUM";
+
+    if (prop_name == "ROCKNUM")
+        return "ROCKNUM";
+
+    throw std::invalid_argument("The rocknum propertype: " + prop_name + " is not valid");
+}
+
+
+RockConfig::Hysteresis hysteris(const std::string& s) {
```
Comment (bska) [inline src/opm/parser/eclipse/EclipseState/SimulationConfig/RockConfig.cpp]: Do you want to rename this function before we merge?

File: src/opm/parser/eclipse/EclipseState/SimulationConfig/RockConfig.cpp [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+
+std::string num_prop(const std::string& prop_name) {
+    if (prop_name == "PVTNUM")
+        return "PVTNUM";
+
+    if (prop_name == "SATNUM")
+        return "SATNUM";
+
+    if (prop_name == "ROCKNUM")
+        return "ROCKNUM";
```
Comment (bska) [inline src/opm/parser/eclipse/EclipseState/SimulationConfig/RockConfig.cpp]: This **could** be written as
```C++
if ((prop_name == "PVTNUM") || (prop_name == "SATNUM") || (prop_name == "ROCKNUM"))
    return prop_name;
```
That way you'd avoid a bit of repetition.

File: src/opm/parser/eclipse/EclipseState/SimulationConfig/RockConfig.cpp (hunk 1 above)
Comment (joakim-hove) [reply src/opm/parser/eclipse/EclipseState/SimulationConfig/RockConfig.cpp]: Yes - done.

File: src/opm/parser/eclipse/EclipseState/SimulationConfig/RockConfig.cpp (hunk 2 above)
Comment (joakim-hove) [reply src/opm/parser/eclipse/EclipseState/SimulationConfig/RockConfig.cpp]: I used a third variety - avoiding repetition yes.

File: src/opm/parser/eclipse/EclipseState/SimulationConfig/RockConfig.cpp [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+#include <opm/parser/eclipse/EclipseState/Grid/FieldPropsManager.hpp>
+
+namespace Opm {
+
+namespace {
+
+
+
+std::string num_prop(const std::string& prop_name) {
+    static const std::unordered_map<std::string> prop_names = {"PVTNUM", "SATNUM", "ROCKNUM"};
```
Comment (bska) [inline src/opm/parser/eclipse/EclipseState/SimulationConfig/RockConfig.cpp]: Did you mean `unordered_set` here?

Comment (joakim-hove) [reply src/opm/parser/eclipse/EclipseState/SimulationConfig/RockConfig.cpp]: dang yes

**Pull request merged**